Revise router template functions

pecameron opened the pull request:
OCP 3.7

PR 12882 refactored the router template functions.
This PR documents the changes.

pecameron commented:
@louyihua @rajatchopra @knobunc PTAL

louyihua commented on install_config/router/customized_haproxy_router.adoc:
```diff
@@ -115,21 +115,24 @@ The *define* action names the file that will contain the processed template.
-argument if the variable cannot be read or is empty.
+|`*processEndpointsForAlias(alias ServiceAliasConfig, svc ServiceUnit, action string) []Endpoint*` |
+Returns the list of valid endpoints. When action is "shuffle", the order of endpoints is randomized.
+|`*env(variable, default ...string) string*` | Tries to get the named environment variable from the pod.
```
Here should be `*env(variable string, defaults ...string) string*`

pecameron replied:
changed

louyihua commented on install_config/router/customized_haproxy_router.adoc:
```diff
@@ -115,21 +115,24 @@ The *define* action names the file that will contain the processed template.
-subdomain for hosts (and paths) with a wildcard policy.
-|`*generateRouteRegexp(hostname, path string, wildcard bool) string*` | Generates a regular expression matching the route
-hosts (and paths). The first argument is the host name, the second is the path,
+|`*firstMatch(s string, allowedValues ...string) bool*` | Compares a given string to a list of allowed
```
The first argument of `firstMatch` is a regular expression.

pecameron replied:
changed

pecameron commented:
@louyihua @knobunc PTAL

louyihua commented:
LGTM

Miciah commented on install_config/router/customized_haproxy_router.adoc:
```diff
@@ -115,21 +115,25 @@ The *define* action names the file that will contain the processed template.
 [cols="2*", options="header"]
 |===
 |Function | Meaning
-|`*endpointsForAlias(alias ServiceAliasConfig, svc ServiceUnit) []Endpoint*` | Returns the list of valid endpoints.
```
Is `endpointsForAlias` considered deprecated now?

pecameron replied:
I don't know. It is not currently used.

Miciah commented on install_config/router/customized_haproxy_router.adoc:
```diff
@@ -115,21 +115,25 @@ The *define* action names the file that will contain the processed template.
+|`*processEndpointsForAlias(alias ServiceAliasConfig, svc ServiceUnit, action string) []Endpoint*` |
+Returns the list of valid endpoints. When action is "shuffle", the order of endpoints is randomized.
+|`*env(variable string, default ...string) string*` | Tries to get the named environment variable from the pod.
+If it is not defined or empty, it returns the optional second argument.
```
This doesn't describe the behaviour when no second argument is given, or when more than two arguments are given.

pecameron replied:
Changed. Otherwise it returns the empty string. Other arguments are ignored.

pecameron commented:
@Miciah I addressed your questions and updated the PR. endpointsForAlias is in the code and there are no comments about it being deprecated. It is not currently used in the haproxy template. There is no telling if any customers are using it.

Miciah commented:
@pecameron, thanks! Did you push the new changes? I don't see them yet.

Regarding `endpointsForAlias`, anybody who is using a custom template who has not rebased recently is using the old function.  In order not to break custom templates, we should either document that the old function is deprecated (so that people with custom templates know to change to the new function) or commit to retaining the old function indefinitely (in which case we should document that).

This is probably not the best place to discuss it though; is the matter being tracked anywhere?

pecameron commented:
@miciah  PTAL

Miciah commented on install_config/router/customized_haproxy_router.adoc:
```diff
@@ -115,21 +115,24 @@ The *define* action names the file that will contain the processed template.
+|`*processEndpointsForAlias(alias ServiceAliasConfig, svc ServiceUnit, action string) []Endpoint*` |
+Returns the list of valid endpoints. When action is "shuffle", the order of endpoints is randomized.
+|`*env(variable, default ...string) string*` | Tries to get the named environment variable from the pod.
+If it is not defined or empty, it returns the optional second argument. Otherwise, it returns an empty string.
```
What do you think of this? "If the value is empty or not defined and one or more optional arguments are provided, `env` returns the first non-empty argument. Otherwise, it returns an empty string."

pecameron replied:
@Miciah In practice there are 0 or 1 optional values. The environment variable either has a default or it doesn't.

bfallonf commented:
Sorry @pecameron , looks like this one escaped us. I'll merge this and cherry-pick to the right branches, and make a quick followup PR if I need to. Thanks!

bfallonf commented:
/cherrypick enterprise-3.11

bfallonf commented:
/cherrypick enterprise-3.10

bfallonf commented:
/cherrypick enterprise-3.9

bfallonf commented:
/cherrypick enterprise-3.7
